### PR TITLE
Migrate IndieAuth to full authorization_code grant

### PR DESCRIPTION
**Pull request by jamietanna**
A number of IndieAuth servers do not support the IndieAuth-only
[Profile URL Response], so to widen our support for IndieAuth servers we
can migrate to a full authorization_code token grant.

[Profile URL Response]: https://indieauth.spec.indieweb.org/#profile-url-response

---

FYI: This has not yet been validated, and we are discussing [on #indieweb-dev](https://chat.indieweb.org/dev/2023-01-30#bottom) whether this is the right call

### CODE REVIEW

**Comment by snarfed** [comment]
Thank you for starting this! I'll defer to you and @aaronpk et al on whether it's the right idea, you all know more than me, but if you all say go for it, I'm in. The main changes we'd want here in that case are to bring back the authorization endpoint support, for backcompat, and unify `discover_authorization_endpoint` and `discover_token_endpoint`.

**Comment by snarfed** [comment]
fwiw I tried it against @dshanske's WordPress plugin just now, with a minor bug fix I'll push in a sec, and it died with this error from the plugin:

```json
{
  "error": "parameter_absent",
  "error_description": "Bearer Token Not Supplied or Server Misconfigured to Not Pass Token. Run diagnostic script in WordPress Admin\n\t\t\t\tIndieAuth Settings Page"
}
```

@dshanske does the plugin not yet support the token endpoint flow?

**Comment by dshanske** [comment]
I'm almost certain I did implement both. Let me check the code

**Comment by jamietanna** [comment]
Thanks, sorted the backcompat + some tweaks with the setup - should be good to try again now @snarfed

File: oauth_dropins/indieauth.py [hunk 1]
```diff
@@ -149,35 +205,61 @@ def button_html(cls, *args, **kwargs):
...
+      code_verifier = state.get('code_verifier') or ''
+      validate_resp = util.requests_post(token_endpoint, data={
+        'grant_type': 'authorization_code',
+        'client_id': INDIEAUTH_CLIENT_ID,
+        'code': code,
+        'redirect_uri': request.base_url,
+        'code_verifier': code_verifier,
+      })
+
+      if validate_resp.status_code // 100 == 2:
```
Comment (snarfed) [inline oauth_dropins/indieauth.py:227]: looks like this onward is the same in both methods, mind pulling it out to the bottom so it's unified?

File: oauth_dropins/indieauth.py [hunk 2]
```diff
@@ -54,6 +55,39 @@ def discover_authorization_endpoint(me, resp=None):
     return auth_endpoint
   return INDIEAUTH_URL
 
+def discover_token_endpoint(me, resp=None):
```
Comment (snarfed) [inline oauth_dropins/indieauth.py:58]: this is the same as `discover_authorization_endpoint` except for the rel value, could you unify them?

**Review by snarfed — CHANGES_REQUESTED** [review]
Thank you for the backcompat! Tested against WordPress (my site), works great.

I left a couple comments for duplicated code that should be unified, but otherwise looks great, I'll happily merge once those are done.

**Comment by jamietanna** [comment]
I prefer the "rule of three" for duplication, but happy to have amended :+1:

**Comment by snarfed** [comment]
> I prefer the "rule of three" for duplication, but happy to have amended 👍

Definitely understood! I'm partial to that too pretty often. Thanks for indulging me here. 😁

**Review by snarfed — APPROVED** [review]
Looks great, thank you! Merging and deploying now, both this app and Bridgy Fed.

**Comment by snarfed** [comment]
Getting this stack trace on Bridgy Fed when I use this to unfollow someone, IndieAuthing against my site, WordPress 6.1.1 with IndieAuth plugin 4.2.1. Reverting for now until we can debug.

Oddly I'm not seeing this on https://oauth-dropins.appspot.com/ , it's working fine against my site with this change. 🤔 

```
Traceback (most recent call last):
...
  File "/layers/google.python.pip/pip/lib/python3.9/site-packages/oauth_dropins/indieauth.py", line 223, in dispatch_request
    return self.finish(indie_auth, state=state)
  File "/workspace/follow.py", line 202, in finish
    follower = Follower.get_by_id(state)
  File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/cloud/ndb/_options.py", line 102, in wrapper
    return wrapped(*pass_args, **kwargs)
...
    raise TypeError(_INVALID_ID_TYPE.format(id_))
TypeError: Key ID must be a string or a number; received {'code_verifier': '...', 'me': 'http://snarfed.org', 'state': 'https://blackrock.city/users/spolsky snarfed.org', 'token_endpoint': 'https://snarfed.org/wp-json/indieauth/1.0/token'}
```

**Comment by snarfed** [comment]
Tracked in https://console.cloud.google.com/errors/detail/CP2i-vS2ivTshgE;time=P30D?project=bridgy-federated (for myself)

**Comment by jamietanna** [comment]
Hmm, any idea which key is null that its trying to index into the list? Not one I saw when testing locally with oauth-dropins

**Comment by snarfed** [comment]
Ah, I see it. When we call `finish` here:

https://github.com/snarfed/oauth-dropins/blob/972d4739233e629cfba6c081a846b558634233bf/oauth_dropins/indieauth.py#L223

...we're supposed to pass the user's `state`, the one they originally passed into `redirect_url`. We did that before with this:

https://github.com/snarfed/oauth-dropins/blob/85c42ef6801b5083e0b9720e3a9b72c9245e8737/oauth_dropins/indieauth.py#L163

...but that line got moved into an `else` in this PR, so it only runs for authorization endpoints, not token endpoints.

https://github.com/snarfed/oauth-dropins/blob/972d4739233e629cfba6c081a846b558634233bf/oauth_dropins/indieauth.py#L207

**Comment by jamietanna** [comment]
I'd literally just spotted that too :joy:

**Pull request merged**